### PR TITLE
[Merged by Bors] - feat(Data/Multiset/Basic): add lemma on `Multiset`

### PR DESCRIPTION
**Pull request by qawbecrdtey**
It used to have a following lemma:
```
s.Disjoint t → s ≤ u → t ≤ u → s + t ≤ u
```
Now it only has the lemma:
```
a ∉ s → a ::ₘ s ≤ t ↔ a ∈ t ∧ s ≤ t
```
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Multiset/Basic.lean [hunk 1]
```diff
@@ -2682,6 +2682,42 @@ theorem disjoint_map_map {f : α → γ} {g : β → γ} {s : Multiset α} {t :
     Disjoint (s.map f) (t.map g) ↔ ∀ a ∈ s, ∀ b ∈ t, f a ≠ g b := by
   simp [Disjoint, @eq_comm _ (f _) (g _)]
 
+lemma notmem_singleton_cons_le [DecidableEq α] {a : α} {s : Multiset α} (h : a ∉ s)
+    {t : Multiset α} (ha : a ∈ t) (hs : s ≤ t) :
+    a ::ₘ s ≤ t := by
+  rw [← cons_erase ha, cons_le_cons_iff]
+  exact (cons_erase ha ▸ (@le_cons_of_not_mem _ _ (t.erase a) _ h)).mp hs
+
+lemma notmem_replicate_add_left_le [DecidableEq α] {a : α} {n : ℕ} {s : Multiset α} (h : a ∉ s)
```
Comment (b-mehta) [inline Mathlib/Data/Multiset/Basic.lean:2691]: These two look like special cases of `add_le_of_disjoint` - I'd prefer to only give that one, and ensure that `s.Disjoint (replicate n a)` is easy to prove in the cases you care about.

File: Mathlib/Data/Multiset/Basic.lean [hunk 2]
```diff
@@ -2682,6 +2682,42 @@ theorem disjoint_map_map {f : α → γ} {g : β → γ} {s : Multiset α} {t :
     Disjoint (s.map f) (t.map g) ↔ ∀ a ∈ s, ∀ b ∈ t, f a ≠ g b := by
   simp [Disjoint, @eq_comm _ (f _) (g _)]
 
+lemma notmem_singleton_cons_le [DecidableEq α] {a : α} {s : Multiset α} (h : a ∉ s)
```
Comment (b-mehta) [inline Mathlib/Data/Multiset/Basic.lean:2685]:
```suggestion
lemma cons_le_of_not_mem [DecidableEq α] {a : α} {s : Multiset α} (h : a ∉ s)
```

File: Mathlib/Data/Multiset/Basic.lean (hunk 1 above)
Comment (qawbecrdtey) [reply Mathlib/Data/Multiset/Basic.lean:2691]: I thought I'd use `replicate` lemmas to prove `add_le_of_disjoint`. Maybe I'll just remove both `replicate` lemmas.

File: Mathlib/Data/Multiset/Basic.lean [hunk 3]
```diff
@@ -2682,6 +2682,27 @@ theorem disjoint_map_map {f : α → γ} {g : β → γ} {s : Multiset α} {t :
...
+    [DecidableEq α] {s t : Multiset α} (h : s.Disjoint t) {u : Multiset α}
+    (hs : s ≤ u) (ht : t ≤ u) :
+    s + t ≤ u := by
+  induction s using Multiset.induction_on generalizing t with
+  | empty => simp only [zero_add, ht]
+  | cons a s ih =>
+    rw [cons_add, le_iff_count]; intro x
+    rcases disjoint_cons_left.mp h with ⟨h₁, h₂⟩
+    by_cases h' : x = a <;> simp [h']
+    · simp [count_eq_zero_of_not_mem h₁]
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Multiset/Basic.lean:2701]: Please use `simp?` to make this more robust

File: Mathlib/Data/Multiset/Basic.lean [hunk 4]
```diff
@@ -2682,6 +2682,28 @@ theorem disjoint_map_map {f : α → γ} {g : β → γ} {s : Multiset α} {t :
     Disjoint (s.map f) (t.map g) ↔ ∀ a ∈ s, ∀ b ∈ t, f a ≠ g b := by
   simp [Disjoint, @eq_comm _ (f _) (g _)]
 
+lemma cons_le_of_not_mem [DecidableEq α] {a : α} {s : Multiset α} (h : a ∉ s)
+    {t : Multiset α} (ha : a ∈ t) (hs : s ≤ t) :
+    a ::ₘ s ≤ t := by
```
Comment (YaelDillies) [inline Mathlib/Data/Multiset/Basic.lean:2687]:
```suggestion
    {t : Multiset α} (ha : a ∈ t) :
    a ::ₘ s ≤ t ↔ s ≤ t := by
```

File: Mathlib/Data/Multiset/Basic.lean [hunk 5]
```diff
@@ -2682,6 +2682,28 @@ theorem disjoint_map_map {f : α → γ} {g : β → γ} {s : Multiset α} {t :
...
+  rw [← cons_erase ha, cons_le_cons_iff]
+  exact (cons_erase ha ▸ (@le_cons_of_not_mem _ _ (t.erase a) _ h)).mp hs
+
+theorem add_le_of_disjoint
+    [DecidableEq α] {s t : Multiset α} (h : s.Disjoint t) {u : Multiset α}
+    (hs : s ≤ u) (ht : t ≤ u) :
+    s + t ≤ u := by
```
Comment (YaelDillies) [inline Mathlib/Data/Multiset/Basic.lean:2694]: Can you show that `s + t = s ⊔ t ↔ s.Disjoint t` instead?

File: Mathlib/Data/Multiset/Basic.lean [hunk 6]
```diff
@@ -2674,6 +2680,9 @@ theorem add_eq_union_iff_disjoint [DecidableEq α] {s t : Multiset α} :
   simp_rw [← inter_eq_zero_iff_disjoint, ext, count_add, count_union, count_inter, count_zero,
     Nat.min_eq_zero_iff, Nat.add_eq_max_iff]
 
+theorem add_eq_sup_iff_disjoint [DecidableEq α] {s t : Multiset α} :
+    s + t = s ⊔ t ↔ Disjoint s t := by rw [← add_eq_union_iff_disjoint, sup_eq_union]
+
```
Comment (qawbecrdtey) [inline Mathlib/Data/Multiset/Basic.lean:2685]:
```suggestion
```
I don't think we will ever need this as `add_eq_sup_iff_disjoint` is equivalent to `add_eq_union_iff_disjoint` due to `sup_eq_union`.

File: Mathlib/Data/Multiset/Basic.lean [hunk 7]
```diff
@@ -528,6 +528,12 @@ theorem le_cons_of_not_mem (m : a ∉ s) : s ≤ a ::ₘ t ↔ s ≤ t := by
     perm_middle.subperm_left.2
       ((subperm_cons _).2 <| ((sublist_or_mem_of_sublist s).resolve_right m₁).subperm)
 
+theorem cons_le_of_not_mem (hs : a ∉ s) (ht : a ∈ t) :
+    a ::ₘ s ≤ t ↔ s ≤ t := by
```
Comment (YaelDillies) [inline Mathlib/Data/Multiset/Basic.lean:532]:
```suggestion
    a ::ₘ s ≤ t ↔ a ∈ t ∧ s ≤ t := by
```
and can you prove from the facts that:
* `a ∉ s ↔ Disjoint {a} s`
* `a ::ₘ s = {a} + s`
* `a ∈ t ↔ {a} ≤ t`
* `{a} ∪ s ≤ t ↔ {a} ≤ t ∧ s ≤ t`

**Review by YaelDillies — COMMENTED** [review]
Not easy enough yet, sorry 😉

File: Mathlib/Data/Multiset/Basic.lean (hunk 7 above)
Comment (YaelDillies) [reply Mathlib/Data/Multiset/Basic.lean:532]: Sorry, also remove `ht : a ∈ t` from the hypotheses

Comment (qawbecrdtey) [reply Mathlib/Data/Multiset/Basic.lean]: `≤` for Multisets have just been declared around line 500, so there's not much lemmas to use around that line. Check the new commit @YaelDillies

Comment (YaelDillies) [reply Mathlib/Data/Multiset/Basic.lean]: > `≤` for Multisets have just been declared around line 500, so there's not much lemmas to use around that line

Why not prove the lemma further down in the file, then?

Comment (qawbecrdtey) [reply Mathlib/Data/Multiset/Basic.lean]: First, there is no need to use more advanced lemmas. It only takes <= 5 lines even without them. Second, L518 declares a similar lemma `le_cons_of_not_mem`, and it is nice for similar lemmas to go together.

**Review by YaelDillies — COMMENTED** [review]
maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**